### PR TITLE
Add subclass CustomCSVLogger to enable tracking additional data.

### PR DESCRIPTION
**Pull request by tx1985**
### Summary
This is an updated PR, following feedback from, and replacing, #12266, in case the discussed feature is useful for others, too.

I use this PR for my own models: Enable inclusion of additional data columns in the csv file from callback `CSVLogger` by introducing a subclass `CustomCSVLogger`. For example, a researcher might want to add the seconds per epoch, an identifier (same for every epoch) and the learning rate of each epoch (especially useful when using a learning rate scheduler):

```
    import numpy as np
    from keras import backend as K

    def get_learning_rate(epoch=None, model=None):
        return np.round(float(K.get_value(model.optimizer.lr)), 5)

    custom_csv_logger = CustomCSVLogger('training.log',
        additional_columns={'seconds': None, 'learning_rate': get_learning_rate})
    model.fit(X_train, Y_train, callbacks=[custom_csv_logger])
```

Parameter `additional_columns`: dictionary with key-value pairs, such that key is the name of the column to add to the csv file. If key is 'seconds' then the value is ignored and the number of seconds per epoch will be printed in the respective column 'seconds'. For other keys the value has to be a callable which must have the following two keyword arguments: `epoch` (type integer) and `model` (type Model).

### Related Issues
- current PR follows feedback from, and replaces, #12266 
- current PR includes feature "ability to add static values to output file" from #4157 

### PR Overview
- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [x] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### Further details
- I have included unit tests at `keras/tests/keras/test_callbacks.py`, which are all passing.
- I have included documentation and a code example at `keras/keras/callbacks.py`
- All code changes that I did are in the above two py-files.
- This PR takes my [API Design Review suggestion](https://docs.google.com/document/d/1oU-x3l7wcZYmuk1bopnxMBs1YMyfSjQvX15k4Yn5hdA/edit#heading=h.5ojovhmssrde) a little bit further, and is a bit more clear, I believe. The latter suggestion was posted in the respective [GoogleGroup](https://groups.google.com/forum/#!topic/keras-users/lFARnrJdEcs) and had just 1 positive comment, but no further discussion.

Looking forward to feedback and improvement suggestions.

### CODE REVIEW

**Comment by fchollet** [comment]
Hi,

To be clear, I was suggesting that since this class is maintainable as a small subclass of the original callback, you could easily maintain it as part of your own code with pushing it into the Keras API :)

Subclassing to extend the API is part of the way the API is intended to be used.

In the Keras API, every new feature has to be maintained in perpetuity, and has to be replicated in every implementation of the Keras API (which includes tf.keras, multi-backend Keras, tensorflow.js, keras-mxnet, and others).

As, such, our criteria for adding a new feature in the API is the following:

- It should be broadly useful to our users, rather than a niche feature. Niche features should be maintained independently by those who need them (e.g. by extending the API via subclassing), as third-party add-on packages.
- It should be widely recognized as a machine learning best practice. We will not add new layers/etc that were recently published to ArXiv.org, even in case of claims of increased accuracy/etc. We only add new objects that are already commonly used in the machine learning community. Presumably, a new technique that does result in meaningful gains would be broadly adopted after a few months anyway (like ResNet), and that’s when we would be adding it to the core API.
- It should have an owner committed to maintaining it in the long term. In particular, the code should be maintainable by multiple people on the team, not just by one technical guru.

**Comment by tx1985** [comment]
Ok, thanks for clarifying :)
